### ISSUE TITLE
sql: query decorrelation error under recursive CTE in asyncpg

### ISSUE THREAD

**Issue by gordthompson** [cockroachdb/cockroach#71908]
tl;dr: https://github.com/cockroachdb/cockroach/issues/71908#issuecomment-951320600

----

**Describe the problem**

I've seen a number of issues regarding "could not decorrelate subquery" (e.g., #68086 , #67951 ) but this one might be a little different in that it seems to be related more specifically to asyncpg.

Reflection for PostgreSQL in SQLAlchemy uses queries that include `ANY()`. For example, to retrieve primary key columns it emits a query similar to this:

```sql
SELECT
cons.conname as name,
cons.conkey as key,
a.attnum as col_num,
a.attname as col_name
FROM
pg_catalog.pg_constraint cons
join pg_attribute a
  on cons.conrelid = a.attrelid AND
    a.attnum = ANY(cons.conkey)
WHERE
cons.contype = 'p'
```

If I create a table in an empty CockroachDB database

```sql
CREATE TABLE tbl1 (id int primary key, txt varchar(50))
```

then this code (using psycopg2) works fine:

```python
import psycopg2

reflect_query = """\
SELECT
cons.conname as name,
cons.conkey as key,
a.attnum as col_num,
a.attname as col_name
FROM
pg_catalog.pg_constraint cons
join pg_attribute a
  on cons.conrelid = a.attrelid AND
    a.attnum = ANY(cons.conkey)
WHERE
cons.contype = 'p'
"""

# connect to CockroachDB
conn = psycopg2.connect(host="localhost", port=26257, user="root", dbname="defaultdb")
cur = conn.cursor()
cur.execute(reflect_query)
print(cur.fetchall())
# [('primary', [1], 1, 'id')]
```

However, if I connect to CockroachDB using asyncpg I get the "could not decorrelate subquery" error

```python
import asyncio
import asyncpg

reflect_query = """\
SELECT
cons.conname as name,
cons.conkey as key,
a.attnum as col_num,
a.attname as col_name
FROM
pg_catalog.pg_constraint cons
join pg_attribute a
  on cons.conrelid = a.attrelid AND
    a.attnum = ANY(cons.conkey)
WHERE
cons.contype = 'p'
"""

async def run():
    cr = {"host": "localhost", "port": 26257, "user": "root", "database": "defaultdb"}
    conn = await asyncpg.connect(**cr)
    values = await conn.fetch(reflect_query)  # Error: "could not decorrelate subquery"
    print(values)
    await conn.close()


loop = asyncio.get_event_loop()
loop.run_until_complete(run())
```

However, if I use asyncpg to connect to PostgreSQL then the same query does not fail:

```python
async def run():
    pg = {"host": "192.168.0.199", "user": "scott", "password": "tiger", "database": "test"}
    conn = await asyncpg.connect(**pg)
    values = await conn.fetch(reflect_query)
    print(values)
    # [<Record name='tbl1_pkey' key=[1] col_num=1 col_name='id'>]
    await conn.close()
```

It also seems to be directly related to `ANY()` because if I remove the `AND a.attnum = ANY(cons.conkey)` then the error does not occur for cockroachdb+asyncpg.

**Environment:**
 - CockroachDB CCL v21.2.0-rc.1 (x86_64-unknown-linux-gnu, built 2021/10/18 14:45:39, go1.16.6)
 - Server OS: Xubuntu 20.04
 - asyncpg 0.24.0 (compared with psycopg2 2.9.1)

**Additional context**

One possible workaround might be to have the sqlalchemy-cockroachdb dialect override SQLAlchemy's code for PostgreSQL reflection with code that does not emit queries that use `ANY()`.

Epic CRDB-18886
Jira issue: CRDB-10828

**Comment by michae2** [comment]
Tagging @vy-ton

**Comment by gnat** [comment]
Any updates on this? @knz Currently a blocker for ARRAY support in Piccolo ORM.

Also affects our M2M "join table" support, since it uses ARRAY.

Example query where we get `asyncpg.exceptions._base.UnknownPostgresError: could not decorrelate subquery`:

```sql
SELECT "band"."name", 
        ARRAY(
            SELECT
                "inner_genre"."name"
            FROM 
"genre_to_band"
JOIN "band" "inner_band" ON (
    "genre_to_band"."band" = "inner_band"."id"
)
JOIN "genre" "inner_genre" ON (
    "genre_to_band"."genre" = "inner_genre"."id"
)
WHERE "genre_to_band"."band" = "band"."id"

        ) AS "genres"
     FROM band
```

**Comment by gnat** [comment]
That's great. We'll move forward with a Cockroach release for Piccolo ORM relatively soon- if we cannot figure something out ourselves, we'll just flag the affected features as in-progress and keep an eye out here.
